### PR TITLE
Add support for multi segment base urls

### LINKED ISSUE #1984: BaseUrl only allows for lowercase and setting it to something breaks Chromecast
**Describe the bug**
You cannot set the BaseUrl to include an uppercase letter otherwise you're unable to login and when you do set the string lowercase then Chromecasting doesn't work.

**Expected behavior**
<BaseUrl>/Suss</BaseUrl>
*Able to login and utilize Chromecast*
**What actually happens**
<BaseUrl>/Wew</BaseUrl>
*Unable to login*
Or
<BaseUrl>[anythinglowercase]</BaseUrl>
*Still unable to Chromecast*

**System:**
 - OS: Arch Linux
 - Browser: Firefox, Chromium
 - Jellyfin Version: 10.4.1-1
 - Reverse proxy: nginx and direct access.

**Additional context**
Again this is running on Arch Linux installed through the AUR.

### PR DESCRIPTION
**Pull request by Bond-009**
`GetPathValue` now return a `ReadOnlySpan<char>`, so it doesn't allocate any new strings.
## Issues
Fixes #1984

### CODE REVIEW

File: MediaBrowser.Api/BaseApiService.cs [hunk 1]
```diff
@@ -16,30 +18,43 @@ namespace MediaBrowser.Api
...
+            ResultFactory = httpResultFactory;
+        }
+
         /// <summary>
-        /// Gets or sets the logger.
+        /// Gets the logger.
         /// </summary>
         /// <value>The logger.</value>
-        public ILogger Logger => ApiEntryPoint.Instance.Logger;
+        protected ILogger Logger { get; }
```
Comment (JustAMan) [inline MediaBrowser.Api/BaseApiService.cs]: why?

File: MediaBrowser.Api/BaseApiService.cs [hunk 2]
```diff
@@ -16,30 +18,43 @@ namespace MediaBrowser.Api
...
+        /// <value>The server configuration manager.</value>
+        protected IServerConfigurationManager ServerConfigurationManager { get; }
 
         /// <summary>
-        /// Gets or sets the HTTP result factory.
+        /// Gets the HTTP result factory.
         /// </summary>
         /// <value>The HTTP result factory.</value>
-        public IHttpResultFactory ResultFactory => ApiEntryPoint.Instance.ResultFactory;
+        protected IHttpResultFactory ResultFactory { get; }
```
Comment (JustAMan) [inline MediaBrowser.Api/BaseApiService.cs]: again, why?

File: MediaBrowser.Api/EnvironmentService.cs [hunk 3]
```diff
@@ -120,13 +122,14 @@ public class EnvironmentService : BaseApiService
...
+            IHttpResultFactory httpResultFactory,
+            INetworkManager networkManager,
+            IFileSystem fileSystem)
+            : base(logger, serverConfigurationManager, httpResultFactory)
         {
-            if (networkManager == null)
-            {
-                throw new ArgumentNullException(nameof(networkManager));
-            }
-
```
Comment (JustAMan) [inline MediaBrowser.Api/EnvironmentService.cs]: why?

File: MediaBrowser.Api/Images/ImageService.cs [hunk 4]
```diff
@@ -440,9 +449,9 @@ public Task Post(PostUserImage request)
         /// <param name="request">The request.</param>
         public Task Post(PostItemImage request)
         {
-            var id = GetPathValue(1);
+            var id = Guid.Parse(GetPathValue(1));
 
-            request.Type = (ImageType)Enum.Parse(typeof(ImageType), GetPathValue(3), true);
+            request.Type = Enum.Parse<ImageType>(GetPathValue(3).ToString(), true);
```
Comment (JustAMan) [inline MediaBrowser.Api/Images/ImageService.cs]: I don't like those hardcoded numbers of where in query is the argument (like `3`), can this be handled more elegantly?

File: MediaBrowser.Api/UserLibrary/ItemsService.cs [hunk 5]
```diff
@@ -58,25 +58,17 @@ public class ItemsService : BaseApiService
...
-            }
-            if (localization == null)
-            {
-                throw new ArgumentNullException(nameof(localization));
-            }
-            if (dtoService == null)
-            {
-                throw new ArgumentNullException(nameof(dtoService));
-            }
-
```
Comment (JustAMan) [inline MediaBrowser.Api/UserLibrary/ItemsService.cs]: why?

File: MediaBrowser.Controller/Library/IUserDataManager.cs [hunk 6]
```diff
@@ -26,13 +26,11 @@ public interface IUserDataManager
         /// <param name="userData">The user data.</param>
         /// <param name="reason">The reason.</param>
         /// <param name="cancellationToken">The cancellation token.</param>
-        /// <returns>Task.</returns>
         void SaveUserData(Guid userId, BaseItem item, UserItemData userData, UserDataSaveReason reason, CancellationToken cancellationToken);
         void SaveUserData(User userId, BaseItem item, UserItemData userData, UserDataSaveReason reason, CancellationToken cancellationToken);
 
         UserItemData GetUserData(User user, BaseItem item);
 
-        UserItemData GetUserData(string userId, BaseItem item);
```
Comment (JustAMan) [inline MediaBrowser.Controller/Library/IUserDataManager.cs]: why did you remove this helper btw?

File: MediaBrowser.Controller/Library/IUserManager.cs [hunk 7]
```diff
@@ -49,20 +49,13 @@ public interface IUserManager
...
         /// <returns>The user with the specified Id, or <c>null</c> if the user doesn't exist.</returns>
         /// <exception cref="ArgumentException"><c>id</c> is an empty Guid.</exception>
         User GetUserById(Guid id);
 
-        /// <summary>
-        /// Gets the user by identifier.
-        /// </summary>
-        /// <param name="id">The identifier.</param>
-        /// <returns>User.</returns>
-        User GetUserById(string id);
```
Comment (JustAMan) [inline MediaBrowser.Controller/Library/IUserManager.cs]: why remove the helper?

File: MediaBrowser.Model/Configuration/ServerConfiguration.cs [hunk 8]
```diff
@@ -172,16 +172,18 @@ public string BaseUrl
...
+
+                if (value[0] != '/')
                 {
                     // If baseUrl was not configured with a leading slash, append one for consistency
                     value = "/" + value;
                 }
 
                 // Normalize the end of the string
-                if (value.EndsWith("/"))
+                if (value[value.Length - 1] == '/')
```
Comment (JustAMan) [inline MediaBrowser.Model/Configuration/ServerConfiguration.cs]: original version was clearer, and I don't think speed matters here so much

File: MediaBrowser.Api/BaseApiService.cs (hunk 1 above)
Comment (JustAMan) [reply MediaBrowser.Api/BaseApiService.cs]: actually, I should have asked a more generic question - why turn this class into an `abstract` one, what's the reasoning?

File: MediaBrowser.Api/ApiEntryPoint.cs [hunk 9]
```diff
@@ -684,178 +668,24 @@ private void DeleteHlsPartialStreamFiles(string outputFilePath)
...
-                return;
-            }
-
-            lock (_timerLock)
-            {
-                if (KillTimer == null)
-                {
-                    Logger.LogDebug("Starting kill timer at {0}ms. JobId {1} PlaySessionId {2}", intervalMs, Id, PlaySessionId);
-                    KillTimer = new Timer(new TimerCallback(callback), this, intervalMs, Timeout.Infinite);
+                    (exs ??= new List<Exception>(4)).Add(ex);
```
[the remark this answers is not in the document]
Comment (Bond-009) [reply MediaBrowser.Api/ApiEntryPoint.cs]: The default size is 0, after that it’s 4, 8, 16 etc.

File: MediaBrowser.Api/BaseApiService.cs (hunk 1 above)
Comment (Bond-009) [reply MediaBrowser.Api/BaseApiService.cs]: It should only be inherited and never created a direct instance of

File: MediaBrowser.Api/EnvironmentService.cs (hunk 3 above)
Comment (Bond-009) [reply MediaBrowser.Api/EnvironmentService.cs]: Can’t be null, gets passed by the ServiceProvider

File: MediaBrowser.Api/Images/ImageService.cs (hunk 4 above)
Comment (Bond-009) [reply MediaBrowser.Api/Images/ImageService.cs]: Maybe, but that’s out of the scope of this PR

File: MediaBrowser.Api/UserLibrary/ItemsService.cs (hunk 5 above)
Comment (Bond-009) [reply MediaBrowser.Api/UserLibrary/ItemsService.cs]: Again, DI makes sure these aren’t null

File: MediaBrowser.Controller/Library/IUserDataManager.cs (hunk 6 above)
Comment (Bond-009) [reply MediaBrowser.Controller/Library/IUserDataManager.cs]: It encourages bad practice.

File: MediaBrowser.Controller/Library/IUserManager.cs (hunk 7 above)
Comment (Bond-009) [reply MediaBrowser.Controller/Library/IUserManager.cs]: Same

File: MediaBrowser.Api/BaseApiService.cs (hunk 1 above)
Comment (JustAMan) [reply MediaBrowser.Api/BaseApiService.cs]: But is that _required_ to be done in this PR? Looks like a general architecture cleanup to me.

File: MediaBrowser.Api/Images/ImageService.cs (hunk 4 above)
Comment (JustAMan) [reply MediaBrowser.Api/Images/ImageService.cs]: that's true...

File: MediaBrowser.Controller/Library/IUserDataManager.cs (hunk 6 above)
Comment (JustAMan) [reply MediaBrowser.Controller/Library/IUserDataManager.cs]: I don't understand why it is so. We're receiving this parameter as a string from the outside world, what's so bad in passing it further in a helper function?

File: MediaBrowser.Api/BaseApiService.cs (hunk 1 above)
Comment (Bond-009) [reply MediaBrowser.Api/BaseApiService.cs]: This logger property change is, the  addition of abstract not.

File: MediaBrowser.Api/BaseApiService.cs (hunk 2 above)
Comment (Bond-009) [reply MediaBrowser.Api/BaseApiService.cs]: Can’t (easily) access the instance of ApiEntryPoint in testing code.

File: MediaBrowser.Controller/Library/IUserDataManager.cs (hunk 6 above)
Comment (Bond-009) [reply MediaBrowser.Controller/Library/IUserDataManager.cs]: It’s a 1 function call to convert it to a guid, and we are only receiving it as a string cause the current api code doesn’t convert it to a guid by itself, ASP.net core can.

File: MediaBrowser.Api/BaseApiService.cs (hunk 1 above)
Comment (JustAMan) [reply MediaBrowser.Api/BaseApiService.cs]: well, if we remove all fields then it doesn't make sense to _not_ make it abstract... so I'll stop objecting here :)

File: Emby.Dlna/Api/DlnaServerService.cs [hunk 10]
```diff
@@ -206,49 +204,103 @@ private ControlResponse PostAsync(Stream requestStream, IUpnpService service)
...
+            static void ThrowIndexOutOfRangeException()
+            {
+                throw new IndexOutOfRangeException("Path doesn't contain enough segments.");
+            }
 
-            string baseUrl = _configurationManager.Configuration.BaseUrl;
+            static void ThrowInvalidDataException()
+            {
+                throw new InvalidDataException("Path doesn't start with the base url.");
+            }
```
[the remark this answers is not in the document]
Comment (Bond-009) [reply Emby.Dlna/Api/DlnaServerService.cs]: It's faster and common practice, if you take a look at the standard library you'll see all (newer) code uses a `ThrowHelper` class to throw exceptions. This is functionally the same.

File: Emby.Dlna/Api/DlnaServerService.cs [hunk 11]
```diff
@@ -206,49 +204,103 @@ private ControlResponse PostAsync(Stream requestStream, IUpnpService service)
...
+            // Remove the query string
+            pos = path.LastIndexOf('?');
+            if (pos != -1)
             {
-                index++;
+                path = path.Slice(0, pos);
             }
-            else if (string.Equals(first, baseUrl.Remove(0, 1)))
+
+            // Remove the domain
```
[the remark this answers is not in the document]
Comment (Bond-009) [reply Emby.Dlna/Api/DlnaServerService.cs]: You *could* use *System.Uri*, but that isn't the most clean/efficient thing IMO

File: Emby.Dlna/Api/DlnaServerService.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (JustAMan) [reply Emby.Dlna/Api/DlnaServerService.cs]: This adds extra lines (as it is used once), and why is it faster?.. just wondering

[the remark this answers is not in the document]
Comment (Bond-009) [reply Emby.Dlna/Api/DlnaServerService.cs]: https://reubenbond.github.io/posts/dotnet-perf-tuning
